lib/class-sp-sync-manager.php

```diff
@@ -29,47 +29,50 @@ class SP_Sync_Manager extends SP_Singleton {
 	public $published_posts = false;
 
 	/**
-	 * Sync a single post (on creation or update)
-	 *
-	 * @todo if post should not be added, it's deleted (to account for unpublishing, etc). Make that more elegant.
-	 *
-	 * @param int $post_id The post ID of the post to be sync'd.
-	 * @access public
+	 * Setup the singleton.
 	 */
-	public function sync_post( $post_id ) {
-		$post     = new SP_Post( get_post( $post_id ) );
-		$response = SP_API()->index_post( $post );
-
-		if ( is_wp_error( $response ) && 'unindexable-post' === $response->get_error_code() ) {
-			// If the post should not be indexed, ensure it's not in the index already.
-			// @todo This is excessive, figure out a better way around it.
-			$this->delete_post( $post_id );
-			do_action( 'sp_debug', "[SP_Sync_Manager] Post {$post_id} is not indexable", $response );
-			return;
+	public function setup() {
+		if ( SP_Config()->active() ) {
+			// When posts & attachments get_updated, queue up syncs.
+			add_action( 'save_post',       array( $this, 'sync_post' ) );
+			add_action( 'edit_attachment', array( $this, 'sync_post' ) );
+			add_action( 'add_attachment',  array( $this, 'sync_post' ) );
+			add_action( 'deleted_post',    array( $this, 'delete_post' ) );
+			add_action( 'trashed_post',    array( $this, 'delete_post' ) );
+
+			// When terms or term relationships get updated, queue up syncs.
+			// add_action( 'added_term_relationship',    array( $this, 'sync_post' ) );
+			// add_action( 'deleted_term_relationships', array( $this, 'sync_post' ) );
+
+			// When users get updated, queue up syncs for their posts.
+			// @TODO
 		}
+	}
 
-		if ( ! $this->parse_error( $response, array( 200, 201 ) ) ) {
-			do_action( 'sp_debug', "[SP_Sync_Manager] Indexed Post {$post_id}", $response );
-		} else {
-			do_action( 'sp_debug', "[SP_Sync_Manager] Error Indexing Post {$post_id}", $response );
-		}
+	/**
+	 * Enqueue a single post to be indexed.
+	 *
+	 * @param int $post_id The post ID of the post to be indexed.
+	 */
+	public function sync_post( $post_id ) {
+		update_post_meta( $post_id, '_sp_index', '1' );
+		SP_Cron()->schedule_queue_index();
 	}
 
 	/**
 	 * Delete a post from the ES index.
 	 *
-	 * @param int $post_id The post ID of the post to delete.
-	 * @access public
+	 * @param int $post_id The post ID of the post to delete from Elasticsearch.
 	 */
 	public function delete_post( $post_id ) {
-		$response = SP_API()->delete_post( $post_id );
-
-		// We're OK with 404 responses here because a post might not be in the index.
-		if ( ! $this->parse_error( $response, array( 200, 404 ) ) ) {
-			do_action( 'sp_debug', '[SP_Sync_Manager] Deleted Post', $response );
-		} else {
-			do_action( 'sp_debug', '[SP_Sync_Manager] Error Deleting Post', $response );
+		$to_delete = get_option( 'sp_delete', array() );
+		if ( ! is_array( $to_delete ) ) {
+			$to_delete = array();
 		}
+		$to_delete[] = absint( $post_id );
+
+		update_option( 'sp_delete', array_unique( $to_delete ), 'no' );
```

Comment:
Potential overwrite race condition here. Need to figure out a better way to store this data if we're to proceed with this route. Maybe deletes should be synchronous, though?